File: flink/src/main/java/com/data/flink/process/datastream/FlinkStreamDemo.java
package com.data.flink.process.datastream;

import org.apache.flink.runtime.state.filesystem.FsStateBackend;
import org.apache.flink.streaming.api.CheckpointingMode;
import org.apache.flink.streaming.api.environment.CheckpointConfig;
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment;

public class FlinkStreamDemo {

    StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

    String checkpoint = "";

    public static void main(String[] args) {
        env.enableCheckpointing();
        env.getCheckpointConfig().setCheckpointingMode(CheckpointingMode.AT_LEAST_ONCE);

        env.getCheckpointConfig().setCheckpointTimeout(60000L);
        env.getCheckpointConfig().setMaxConcurrentCheckpoints(1);
        env.getCheckpointConfig().enableExternalizedCheckpoints(CheckpointConfig.ExternalizedCheckpointCleanup.RETAIN_ON_CANCELLATION);

        env.setStateBackend(new FsStateBackend(checkpint));


    }
}

File: flink/src/main/scala/com/data/flink/process/dataset/FlinkStreamDemo.scala
package com.data.flink.process.dataset

import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.api.windowing.time.Time

object FlinkStreamDemo {
  def main(args: Array[String]): Unit = {

    val env = StreamExecutionEnvironment.getExecutionEnvironment

    val dataStream: DataStream[String] = env.socketTextStream("127.0.0.1",9000)

    dataStream.flatMap{ line => line.toLowerCase().split(",")}
      .filter(_.nonEmpty)
      .map {word => (word, 1)}
      .keyBy(0)
      .timeWindow(Time.seconds(3))
      .sum(1)
      .print()

    env.execute("streaming wordcount")

  }

}
